### PR TITLE
Add auto cluster configuration support

### LINKED ISSUE #792: Auto Clustering of nodes should be enabled
Hello, 

I know there was alot of discuss on this project regarding adding rabbitmq clustering feature. 
at the moment, when i use this module and run puppet on two different nodes with same erlang cookie, one after the other.  i am getting two different instances of rabbitmq but not a cluster instance. 

https://github.com/voxpupuli/puppet-rabbitmq/pull/718

## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet:
- Ruby:
- Distribution:
- Module version:
```puppet
#################################################################################
## This class will install and configure Rabbitmq Cluster with neccessay repo.###
#################################################################################
class role::linux::rabbitmq::rabbitmqinstall (

  $package_ensure   = $role::linux::rabbitmq::rmqparams::package_ensure,
  $package_name     = $role::linux::rabbitmq::rmqparams::package_name,
  $package_source   = $role::linux::rabbitmq::rmqparams::package_source,
  $package_erlang   = $role::linux::rabbitmq::rmqparams::package_erlang,
  $version_erlang   = $role::linux::rabbitmq::rmqparams::version_erlang,
  $cluster_nodes = $role::linux::rabbitmq::rmqparams::cluster_nodes,
  $config_cluster    = true,
  $cluster_node_type = $role::linux::rabbitmq::rmqparams::cluster_node_type,
  $port              = $role::linux::rabbitmq::rmqparams::port,
  $erlang_cookie     = 'VKCRHNHTENXABMBFFTHQ',
  $wipe_db_on_cookie_change = $role::linux::rabbitmq::rmqparams::wipe_db_on_cookie_change,
  $plugin_enable  = $role::linux::rabbitmq::rmqparams::plugin_enable,
  $rabbitmq_home  = $role::linux::rabbitmq::rmqparams::rabbitmq_home,
  $rabbitmq_user  = $role::linux::rabbitmq::rmqparams::rabbitmq_user,
  $rabbitmq_group = $role::linux::rabbitmq::rmqparams::rabbitmq_group,
  $clustername    = $role::linux::rabbitmq::rmqparams::clustername,
  $host           = $facts['hostname'],
  $pluginname  = lookup ({'name'  => 'rabbitmq::plugin','merge'  => 'unique', default_value  => {}})

) inherits role::linux::rabbitmq::rmqparams {
  include role::linux::rabbitmq::rabbitmqrepo
  include role::linux::rabbitmq::rabbitmqerlang

## Below code ensure package erlang dependancy is installed before Rabbitmq
notify {"To install erlang version ${version_erlang}":}
    package { $package_erlang:
        ensure  => $version_erlang,
        name    => $package_erlang,
        require => Class['role::linux::rabbitmq::rabbitmqerlang'],
        before  => Class['rabbitmq'],
      }

## Below code will add the repo for rabbitmq ,install Rabbitmq server 3.7.7 and enable the service.
  notify { "To install Rabbitmq version ${package_ensure}":}
  notify { "To install Rabbitmq erlang ${erlang_cookie}":}
  if ($config_cluster == true) {
      #fail('I am sorry i dont know what to now')
    class { 'rabbitmq':
      package_ensure           => $package_ensure,
      package_name             => 'rabbitmq-server',
      config_cluster           => $config_cluster,
      cluster_nodes            => $cluster_nodes,
      cluster_node_type        => $cluster_node_type,
      delete_guest_user        => true,
      erlang_cookie            => $erlang_cookie,
      wipe_db_on_cookie_change => $wipe_db_on_cookie_change,
      require                  =>[
        Package[$package_erlang],
        Class['role::linux::rabbitmq::rabbitmqrepo'],
      ],
    }
  } else {
      fail('Node is not part of the cluster')
    }
  notify{" Cluster nodes are ${cluster_nodes}":}
  notify {"Following plugins will be enabled ${pluginname}":}
  if ($plugin_enable == true) {
    $pluginname.each |$namevar| {
      rabbitmq_plugin { $namevar:
        ensure   => present,
        provider => 'rabbitmqplugins',
      }
    }
  }
}
```

## What are you seeing

Two different rabbitmq instance. 

## What behaviour did you expect instead

a cluster rabbitmq instance.

### PR DESCRIPTION
**Pull request by fatpat**
#### Pull Request (PR) description
This PR add the clustering management
- added clustername fact and spec tests for it
- added rabbitmq_cluster type and provider with tests
- rework of PR #736 

#### This Pull Request (PR) fixes the following issues
Fixes #792
Closes #736

### CODE REVIEW

**Comment by fatpat** [comment]
Still working on tests...

**Comment by fatpat** [comment]
tests are now all green. Only one is failing due to network issues.

it's good to me

@wyardley @bastelfreak

**Comment by fatpat** [comment]
it would be great to include this PR in the coming release (Release 11.1.0 #879)

**Comment by wyardley** [comment]
@fatpat have you tested this with an actual auto-cluster setup as well?

Only other question is, it seems like the original author still has some credit from the commit email; if this is based on one or more previous PRs, is it worth also mentioning the authors in the commit message?

File: examples/cluster/join_cluster_and_change_name.pp [hunk 1]
```diff
@@ -0,0 +1,4 @@
+rabbitmq_cluster { 'test_cluster':
+  init_node => 'host1',
+}
+# This will join host2 to host1s cluster and set the cluster name to test_cluster
```
Comment (wyardley) [inline examples/cluster/join_cluster_and_change_name.pp:4]: nit, but would it make sense to put the comment above the example code vs. below?

Comment (fatpat) [reply examples/cluster/join_cluster_and_change_name.pp]: fixed + update REFERENCE.md

**Comment by fatpat** [comment]
> @fatpat have you tested this with an actual auto-cluster setup as well?
yes this has been tested and we used it in production.

> 
> Only other question is, it seems like the original author still has some credit from the commit email; if this is based on one or more previous PRs, is it worth also mentioning the authors in the commit message?
fixed in the commit message. My mistake

File: examples/cluster/join_cluster_and_change_name.pp (hunk 1 above)
Comment (wyardley) [reply examples/cluster/join_cluster_and_change_name.pp]: Thanks! So, typically, we'd build reference.md using the rake task as part of the release process, but I can re-run it again (and deal with conflicts that come up) after merging this one.

I had been debating releasing 11.1 separately and having this ship in 11.2, but I'm Ok with putting this in 11.1 if you're confident about it.

**Review by wyardley — APPROVED** [review]

**Pull request merged**